rfc/rfc-0000-simplify-runnable-use.md

```diff
@@ -0,0 +1,318 @@
+                 spec:
+                    accessModes:
+                       - ReadWriteOnce
+                    resources:
+                       requests:
+                          storage: 256Mi
+   urlPath: spec.params[?(@.name=="repository")].value    # <== Refers to path on the immutable object that will be created
+   revisionPath: spec.params[?(@.name=="revision")].value
+   runnable: true                                         # <== Flag
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+When Cartographer processes a template with the runnable flag, rather than 
+creating/updating the templated object it will create/update a Runnable and 
+a ClusterRunTemplate.
+
+The templated object will have two types of fields: hard-coded and templated.
+Hard coded fields can simply be copied directly into the ClusterRunTemplate. 
+Templated fields will be translated. Rather than reading from the specified 
+location, these will read from the runnable's inputs field. This can be done 
+simply by prepending the template paths with runnable.spec.inputs
+
+Example:
+```yaml
+kind: ClusterSourceTemplate    # <== Now a source template
+spec:
+   template:
+         params:
+            - name: repository
+              value: $(workload.spec.source.git.url)$
+```
+
+Will translate to a ClusterRunTemplate with:
+```yaml
+kind: ClusterRunTemplate
+spec:
+   template:
+         params:
+            - name: repository
+              value: $(runnable.spec.inputs.workload.spec.source.git.url)$
+```
+
+and a Runnable with
+```yaml
+kind: Runnable
+spec:
+   inputs:
+      workload.spec.source.git.url: *some-value-from-workload.spec.source.git.url*
+```
+
+The runnable and ClusterRunTemplate will be created with matching names and 
+a matching runTemplateRef. Assuming the example above and a workload with 
+the name `app`:
+
+```yaml
+---
+kind: Runnable
+metadata:
+   name: app-linter
+spec:
+   runTemplateRef:
+      name: app-linter
+---
+kind: ClusterRunTemplate
+metadata:
+   name: app-linter
+```
+
+The ClusterRunTemplate will specify outputs that match the output paths on 
+the template. Given the example above:
+
+```yaml
+---
+kind: ClusterRunTemplate
+spec:
+   outputs:
+      url: spec.params[?(@.name=="repository")].value
+      revision: spec.params[?(@.name=="revision")].value
+```
+
+When reading the runnable, Cartographer will look at the fields that match 
+the type of template. E.g. given a ClusterSourceTemplate with a 
+`runnable:true` field, Cartographer will always create a Runnable and always 
+read the runnable's .status.outputs.url field and .status.outputs.revision 
+field. Similarly for image, carto will read .status.outputs.image; for 
+config it will read .status.outputs.config.
+
+# Migration
+[migration]: #migration
+
+This is an additive change and does not present any breaking changes. Users 
+who wish to continue to hand create ClusterRunTemplates and template out 
+Runnables will be able to do so.
+
+# Drawbacks
```

Comment:
> logically it's no longer an optional, semi-independent component.

How so? The proposal does not change any existing behavior. Users are free to continue writing Runnables/ClusterRunTemplates directly and even to use them outside the context of a supply chain.

Reply:
The workload/supply chain reconciler now has direct knowledge of Runnable. Previously, it had no more knowledge of Runnable than any other resource defined by a template.

Reply:
The workload/supply chain reconciler now has direct knowledge of *some* Runnables. But Runnable remains a component that can be used outside of the context of Cartographer. And a user could still choose to use the current workflow to use Runnable in a supply chain (e.g. the user creates a ClusterRunTemplate and then templates a Runnable within a supply chain template). If the user does so, the workload/supply chain reconciler will have no special knowledge of that Runnable.

Reply:
We have different definitions of what "direct knowledge" means.

Today, we could take the Runnable resource and split it off into an optional extension. The Cartographer core would run just fine without it. After this change, if the Runnable resource was removed, Cartographer itself would break. Cartographer never defines and creates a Runnable resource today, after this change it will.

That's what I mean by "direct knowledge". There's no such thing as partial direct knowledge, either Cartographer knows about the type and how to create it or it doesn't. A template defining how to stamp out a Runnable is not direct knowledge from the perspective of Cartographer, its knowledge from the template.

Reply:
Ah! I see what you mean.

Runnable would be able to work separate from Supply Chains, but Supply Chains would be forever coupled to Runnable. I'm happy to write this into the drawbacks. I'm also happy to take on this dependency. As I stated in the RFC, enabling testing is a prime use case for Cartographer. While a Cartographer user could theoretically depend on some other Tekton wrapper instead of Runnable, in practice Carto supply chain users that want to leverage Tekton have to depend on Runnable.